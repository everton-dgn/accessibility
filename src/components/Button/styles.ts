import styled, { css } from 'styled-components'
import { T } from 'theme'
import { BtnProps } from '.'

const SizesBtn = {
  small: css`
    ${({ theme }) => css`
      height: min-content;
      font-size: ${theme.font.sizes.xSmall};
      padding: ${theme.spacings.xSmall} ${theme.spacings.small};
    `}
  `,
  medium: css`
    ${({ theme }) => css`
      height: min-content;
      font-size: ${theme.font.sizes.small};
      padding: ${theme.spacings.small} ${theme.spacings.xLarge};
    `}
  `,
  large: css`
    ${({ theme }) => css`
      height: min-content;
      font-size: ${theme.font.sizes.small};
      padding: ${theme.spacings.medium} ${theme.spacings.xxxLarge};
    `}
  `,
  xLarge: css`
    ${({ theme }) => css`
      height: min-content;
      font-size: ${theme.font.sizes.small};
      padding: ${theme.spacings.medium} ${theme.spacings.xtreme};
    `}
  `
}

const ColorsBtn = {
  gray: css`
    ${({ theme }) => css`
      background-color: ${theme.colors.gray};

      &:focus {
        box-shadow: 0 0 0 0.2rem #aaaaaa, 0 0.1rem 1.5rem #3e396b80;
      }

      &:hover {
        background-color: ${theme.colors.grayHover};
      }
    `}
  `,
  blue: css`
    ${({ theme }) => css`
      background-color: ${theme.colors.primary};

      &:focus {
        box-shadow: 0 0 0 0.2rem #5872ff, 0 0.1rem 1.5rem #3e396b80;
      }

      &:hover {
        background-color: ${theme.colors.primaryHover};
      }
    `}
  `,
  red: css`
    ${({ theme }) => css`
      background-color: ${theme.colors.red};

      &:focus {
        box-shadow: 0 0 0 0.2rem #fd6e6e, 0 0.1rem 1.5rem #3e396b80;
      }

      &:hover {
        background-color: ${theme.colors.redHover};
      }
    `}
  `
}

export const Container = styled.button<BtnProps>`
  ${({ theme, fullWidth, color, size }) => css`
    cursor: pointer;
    border-radius: ${theme.border.radius};
    font-weight: 400;
    display: inline-flex;
    align-items: center;
    justify-content: center;
    flex-wrap: nowrap;
    border: 0.1rem solid transparent;
    text-shadow: 0 0 0.3rem rgb(0 0 0 / 60%);
    transition: background-color 0.15s ease-in-out;
    width: 100%;
    ${!!color && ColorsBtn[color]};
    ${!!size && SizesBtn[size]}

    ${T.breakPoint(480)} {
      width: ${fullWidth ? '100%' : 'fit-content'};
    }

    &:active {
      box-shadow: none;
    }

    &:disabled,
    &:disabled:hover {
      opacity: 0.5;
      cursor: not-allowed;
      background-color: ${theme.colors.disabled};
    }

    span {
      color: ${theme.colors.white};
    }

    svg + span {
      margin-left: 1rem;
    }

    svg {
      width: 2rem;
      min-width: 2rem;
      height: 2rem;
      min-height: 2rem;
      color: ${theme.colors.white};
    }
  `}
`
