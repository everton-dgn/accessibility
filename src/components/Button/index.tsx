import { ButtonHTMLAttributes, forwardRef, ReactNode } from 'react'
import * as S from 'components/Button/styles'

export type SizeProps = 'small' | 'medium' | 'large' | 'xLarge'

export type ColorProps = 'gray' | 'blue' | 'red'

export interface BtnProps extends ButtonHTMLAttributes<HTMLButtonElement> {
  icon?: ReactNode
  text?: string
  size: SizeProps
  color: ColorProps
  fullWidth?: boolean
}

const Button = forwardRef<HTMLButtonElement, BtnProps>(
  ({ icon, text, ...props }, ref) => (
    <S.Container {...props} ref={ref}>
      {icon}
      {!!text && <span>{text}</span>}
    </S.Container>
  )
)

export default Button
