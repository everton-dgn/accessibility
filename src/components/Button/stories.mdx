import { Canvas, Meta, Story } from '@storybook/addon-docs'
import Button from '.'
import * as stories from './stories'

<Meta
  title="Design System/Button"
  component={Button}
  args={{
    text: 'Clique aqui',
    size: 'large',
    color: 'blue',
    disabled: false,
    fullWidth: false
  }}
  argTypes={{
    size: {
      options: ['small', 'medium', 'large'],
      control: { type: 'inline-radio' }
    },
    color: {
      options: ['gray', 'blue'],
      control: { type: 'inline-radio' }
    },
    onClick: { action: 'onClick' }
  }}
  parameters={{ controls: { exclude: ['icon'], sort: 'requiredFirst' } }}
/>

# Button

This is a general purpose button component.

### Default

The button receives props for styling, icons, as well as a click event,
and makes its reference available through the forwardRef() hook.

<Canvas>
  <Story story={stories.Default} />
</Canvas>
