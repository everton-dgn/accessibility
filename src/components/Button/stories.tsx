import { Story } from '@storybook/react'
import Button, { BtnProps } from '.'

export const Template: Story<BtnProps> = args => <Button {...args} />

export const Default = Template.bind({})
