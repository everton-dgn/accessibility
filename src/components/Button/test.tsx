import { screen } from '@testing-library/react'
import userEvent from '@testing-library/user-event'
import { renderWithTheme } from 'utils/tests/helpers'
import Button from '.'

describe('<Button />', () => {
  it('should render a small blue button with text', () => {
    const { container } = renderWithTheme(
      <Button text="Mais Informações" color="blue" size="small" />
    )

    const btn = screen.getByRole('button', { name: /Mais Informações/i })
    expect(btn).toHaveStyle({
      height: 'min-content',
      padding: '0.8rem 1.2rem',
      fontSize: '1.2rem',
      backgroundColor: '#334BC8'
    })

    expect(btn.closest('button')).not.toBeDisabled()
    expect(container).toMatchSnapshot()
  })

  it('should render a medium blue button with text', () => {
    renderWithTheme(
      <Button size="medium" text="Mais Informações" color="blue" />
    )

    const btn = screen.getByRole('button', { name: /Mais Informações/i })

    expect(btn).toHaveStyle({
      height: 'min-content',
      padding: '1.2rem 2.4rem',
      fontSize: '1.4rem',
      backgroundColor: '#334BC8'
    })
  })

  it('should render a disabled Button', () => {
    renderWithTheme(
      <Button disabled text="Mais Informações" color="blue" size="large" />
    )

    const btn = screen.getByRole('button', { name: /Mais Informações/i })

    expect(btn).toHaveStyle({
      cursor: 'not-allowed',
      opacity: 0.5,
      backgroundColor: '#a1a4b0'
    })
    expect(btn.closest('button')).toBeDisabled()
  })

  it('should call a function once on click on the button', () => {
    const onClick = jest.fn()

    renderWithTheme(
      <Button
        onClick={onClick}
        text="Mais Informações"
        color="blue"
        size="large"
      />
    )

    const btn = screen.getByRole('button', { name: /Mais Informações/i })

    userEvent.click(btn)

    expect(onClick).toHaveBeenCalled()
    expect(onClick).toHaveBeenCalledTimes(1)
  })
})
