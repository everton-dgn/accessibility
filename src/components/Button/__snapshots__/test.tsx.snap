// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Button /> should render a small blue button with text 1`] = `
.c0 {
  cursor: pointer;
  border-radius: 0.4rem;
  font-weight: 400;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-flex-wrap: nowrap;
  -ms-flex-wrap: nowrap;
  flex-wrap: nowrap;
  border: 2px solid transparent;
  text-shadow: 0 0 0.3rem rgb(0 0 0 / 60%);
  -webkit-transition: background-color 0.15s ease-in-out;
  transition: background-color 0.15s ease-in-out;
  width: 100%;
  background-color: #334BC8;
  height: -webkit-min-content;
  height: -moz-min-content;
  height: min-content;
  font-size: 1.2rem;
  padding: 0.8rem 1.2rem;
}

.c0:hover {
  background-color: #5872ff;
}

.c0:disabled,
.c0:disabled:hover {
  opacity: 0.5;
  cursor: not-allowed;
  background-color: #a1a4b0;
}

.c0:focus {
  border: 2px solid #fff;
  box-shadow: 0 0 0 2px #222;
}

.c0 span {
  color: #fff;
}

.c0 svg + span {
  margin-left: 1rem;
}

.c0 svg {
  width: 2rem;
  min-width: 2rem;
  height: 2rem;
  min-height: 2rem;
  color: #fff;
}

@media screen and (min-width:480px) {
  .c0 {
    width: -webkit-fit-content;
    width: -moz-fit-content;
    width: fit-content;
  }
}

<div>
  <button
    class="c0"
    color="blue"
  >
    <span>
      Mais Informações
    </span>
  </button>
</div>
`;
